src/openfermion/ops/_fermion_operator.py

```diff
@@ -205,3 +205,75 @@ def is_molecular_term(self):
             if not (particles == spin == 0):
                 return False
         return True
+
+    def freeze_orbitals(self, occupied, unoccupied=[], prune=True):
+        """Fix some orbitals to be occupied and others unoccupied.
+
+        Removes all operators acting on the specified orbitals, and flips the
+        sign of each term as appropriate. Terms whose expectation value
+        between states with the specified occupancy is zero are removed.
+        Additionally, the indices are renumbered such that the frozen orbitals
+        are eliminated.
+
+        Note that the sign change accounts for both the reordering of operators
+        needed to group the operators acting on the frozen orbitals, as well
+        as the impact of pruning frozen orbitals on the Jordan-Wigner strings
+        of operators acting on unfrozen orbitals.
+
+        Args:
+        occupied: A list containing the indices of the orbitals that are to be
+            assumed to be occupied.
+        unoccupied: A list containing the indices of the orbitals that are to
+            be assumed to be unoccupied.
+        """
+
+        frozen = [(index, 1) for index in occupied] + \
+                 [(index, 0) for index in unoccupied]
+
+        # Loop over each orbital to be frozen. Within each term, move all
+        # ops acting on that orbital to the right side of the term, keeping
+        # track of sign flips that come from swapping operators.
+        for item in frozen:
+            new_terms = {}
+            for term in self.terms:
+                new_term = []
+                new_coef = self.terms[term]
+                current_occupancy = item[1]
+                n_ops = 0  # Number of operations on index that have been moved
+                n_swaps = 0  # Number of swaps that have been done
+
+                for op in enumerate(reversed(term)):
+                    if op[1][0] is item[0]:
+                        n_ops += 1
+
+                        # Determine number of swaps needed to bring the op in
+                        # front of all ops acting on other indices
+                        n_swaps += op[0] - n_ops
+
+                        # Check if the op annihilates the current state
+                        if current_occupancy == op[1][1]:
+                            new_coef = 0
+
+                        # Update current state
+                        current_occupancy = (current_occupancy + 1) % 2
+                    else:
+                        new_term.insert(0, op[1])
+                if n_swaps % 2 is 1:
+                    new_coef *= -1
+                if new_coef is not 0 and current_occupancy is item[1]:
+                    if tuple(new_term) in new_terms:
+                        new_terms[tuple(new_term)] += new_coef
+                    else:
+                        new_terms[tuple(new_term)] = new_coef
+            self.terms = new_terms
```

Comment:
If you are going to create a new copy internally anyway, I think the the default behavior of the function should be to return that new copy and leave the acted upon fermion operator unchanged.  This gives the user the option of overwriting their previous operator, but they don;t have to manually make a deep copy of the whole object if they want to keep the old one around.

Reply:
freeze_orbitals and prune_unused_indices are now functions that return a new copy.